### PR TITLE
Make OpKernelContext as an interface class.

### PR DESCRIPTION
**Pull request by weixingzhang**
Similar to PR#442, the idea of this PR is to make CPU kernels easily leveraged by graph optimization.

For some graph optimizations (e.g. conv+bn fusion, constant folding), the computation needs to be done on model parameters to generate new parameters. To achieve it, the better way is to utilize CPU kernels rather than creating new compute functions. 

At present, the instance of OpKernelContext needs to be created to invoke the kernel. OpKernelContext creation depends on ExecutionFrame. From optimizer point of view, optimizer is not necessary to depend on ExecutionFrame. Instead, optimizer can have a lighter implementation of OpKernelContext. Thus, making OpKernelContext as a pure class and then framework and optimizer both have their own implementation.

### CODE REVIEW

File: include/onnxruntime/core/framework/op_kernel.h [hunk 1]
```diff
@@ -94,80 +114,55 @@ class OpKernelContext {
...
- protected:
-  onnxruntime::NodeIndex GetNodeIndex() const;
-  const SessionState& GetSessionState() const;
+  virtual Fence_t OutputFence(int index) const = 0;
 
-  const MLValue* GetInputMLValue(int index) const;
-  const MLValue* GetImplicitInputMLValue(int index) const;
-  MLValue* GetOutputMLValue(int index);
+  virtual const SessionState* SubgraphSessionState(const std::string& attribute_name) = 0;
+  virtual std::unordered_map<std::string, const MLValue*> GetImplicitInputs() const = 0;
```
Comment (skottmckay) [inline include/onnxruntime/core/framework/op_kernel.h]: Various thing here were private and protected so it was clear they weren't part of the public interface. Not sure that we want to make all those pieces public as some are for purely internal usage (e.g. the subgraph session state). 

Would it be better to instead extract an IOpKernelContext interface with just the basics, and keep the existing OpKernelContext andOpKernelContextInternal implementations as is, with OpKernelContext implementing IOpKernelContext?

It may be possible to merge what was OpKernelContext and OpKernelContextInternal, however Pranav wanted those kept separate.

Comment (kkaranasos) [reply include/onnxruntime/core/framework/op_kernel.h]: I feel that there is some sub-classing needed in the ExecutionFrame.
At the moment the ExecutionFrame and the newly introduced GraphOptimizer have many commonalities in their API.
Could we avoid sub-classing the OpKernelContext altogether and instead have a more lightweight ExecutionFrame to be used only for CPU execution?
If not, maybe we should create an interface for OpKernelContext and one for the ExecutionFrame.
Also, we might want to give another name to the GraphOptimizer (it is essentially a lightweight CPU ExecutionFrame).

Comment (weixingzhang) [reply include/onnxruntime/core/framework/op_kernel.h]: @kkaranasos This PR is to make optimizer leverage kernels without less dependency. With sub-class of ExecutionFrame, that means optimizer will depend on both OpKernelContext and ExecutionFrame, which is not a good idea. And, the APIs in newly introduced GraphOptimizer are much less than ExecutionFrame.

Comment (kkaranasos) [reply include/onnxruntime/core/framework/op_kernel.h]: I don't see the dependency problem. It is not that we are removing dependency to a project (in all cases we have to depend on the framework project).
In the current PR, we have an OpKernelContext that relies on ExecutionFrame and an OpKernelContextOptimizer that relies on GraphOptimizer. All methods of GraphOptimizer are included in ExecutionFrame, so I think the GraphOptimizer is a lightweight ExecutionFrame. So to me it seems natural to share an interface between GraphOptimizer and ExecutionFrame.
Basically it seems that something needs to be organized better, because I see that in the OpKernelContext, there is only one API call of ExecutionFrame that does not also appear in the GraphOptimizer (GetSessionState) -- all the rest are there.

Comment (skottmckay) [reply include/onnxruntime/core/framework/op_kernel.h]: The original description said the PR was about removing the dependency on OpKernelContext because that required ExecutionFrame. However you seen to have duplicated most of the ExecutionFrame logic in GraphOptimizer. Due to that, I'm not quite understanding the overall set of changes.

Why can't you just use ExecutionFrame? 

I want to move the allocation/memory pattern logic outside of ExecutionFrame to simplify it (probably involves creating an interface for handling allocations and passing an implementation of that to the ExecutionFrame ctor). If that happens, would it be usable as is (or as it will soon be - see https://github.com/Microsoft/onnxruntime/pull/470 for some other interface changes that are coming up)?

Comment (weixingzhang) [reply include/onnxruntime/core/framework/op_kernel.h]: @kkaranasos optimizer will be used by converter as a separate library. So, it is better to make it less dependency. Although it depends on framework project now, it doesn't mean it should depend on all codes of framework.  Moreover, framework project could be refactored in the future.

Comment (weixingzhang) [reply include/onnxruntime/core/framework/op_kernel.h]: @skottmckay The purpose of the overall set of changes is to have a separate optimizer which can be shared both by onnx runtime and converter. Also, the CPU kernels can be leveraged by optimizer to compute new model parameters. So, from dependency point of view, the dependency should be as less as possible. 

Why not use ExecutionFrame? The consideration is simple and less dependency. The input of kernel Compute() function is OpKernelContext. So, the intuitive way is to make OpKernelContext as an interface class and optimizer has its own implementation. From optimizer point of view, ExecutionFrame is internal to framework. Moreover, current ExecutionFrame is a little bit complicate. For example, it depends on SessionState, memory planner, etc.

File: include/onnxruntime/core/framework/ml_value.h [hunk 2]
```diff
@@ -69,4 +69,31 @@ class MLValue {
...
+      : tensor_shape{shape} {}
+
+  const TensorShape& GetTensorShape() const {
+    static const TensorShape s_empty_tensor_shape;
+    return tensor_shape != nullptr ? *tensor_shape : s_empty_tensor_shape;
+  }
+
+ private:
+  const TensorShape* tensor_shape{};
+  // todo: is there any parameter needed for ml types?
```
Comment (linkerzhang) [inline include/onnxruntime/core/framework/ml_value.h]: as said, looks to me there's no, if agreed, remove this struct.

File: include/onnxruntime/core/framework/op_kernel.h (hunk 1 above)
Comment (skottmckay) [reply include/onnxruntime/core/framework/op_kernel.h]: There is too much cut-and-pasted code in the PR so I think the current approach needs tweaking.

I think Konstantinos had the right idea to also create an interface for, or simplify the base version of ExecutionFrame.

I want to move all the allocation related things out of the public ExecutionFrame interface anyway as they're internal details and not necessary for its usage in OpKernelContext. That would reduce the public interface to just the methods to get an MLValue, and to get the node offset, and the base implementation should be simple but re-usable. A derived class can have the more complicated stuff using the memory planner etc. 

Would that work? As I wanted to cleanup ExecutionFrame already, I'm happy to do that work and have already spent some time looking into how to do it.

We can see if changes to OpKernelContext are needed after that. If they are, all the methods that were protected in OpKernelContext should probably not be in the base interface. They're all for internal usage in control flow ops only, and this optimizer setup wouldn't support running those ops as currently proposed.

File: include/onnxruntime/core/framework/op_kernel.h [hunk 3]
```diff
@@ -61,21 +61,41 @@ class OpKernelContext {
...
+  }
+
   /**
   Return the number of inputs for a variadic argument.
   @param arg_num The operator argument number.
   @returns Number of inputs the argument has.
   */
-  int NumVariadicInputs(size_t arg_num) const;
+  int NumVariadicInputs(size_t arg_num) const {
+    auto& arg_counts = kernel_->Node().InputArgCount();
```
Comment (linkerzhang) [inline include/onnxruntime/core/framework/op_kernel.h]: QQ: any reason to move the implementation in header file?

File: include/onnxruntime/core/framework/op_kernel.h [hunk 4]
```diff
@@ -94,80 +114,55 @@ class OpKernelContext {
...
-  Fence_t ImplicitInputFence(int index) const;
+  virtual Fence_t ImplicitInputFence(int index) const = 0;
 
   /**
   Return the fence of current node's output identifed by index.
   @param index The index of the output.
   @returns Point to the Fence of the output MLValue.
   It is null if the output MLValue doesn't have fence or the output is optional.
   */
-  Fence_t OutputFence(int index) const;
```
Comment (linkerzhang) [inline include/onnxruntime/core/framework/op_kernel.h]: I think the change of this file should be changing some APIs to allow overwriting, right? if yes, you may keep the original functions' order so that reviewing will be easier :)

Comment (linkerzhang) [reply include/onnxruntime/core/framework/op_kernel.h]: or @weixingzhang you may want to create a pure interface and have this OpKernelContext not changed but one implementation of your interface.

File: include/onnxruntime/core/framework/ml_value.h (hunk 2 above)
Comment (skottmckay) [reply include/onnxruntime/core/framework/ml_value.h]: Totally agree. It adds no value.

File: onnxruntime/core/optimizer/graph_optimizer.cc [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+#include "core/framework/kernel_registry.h"
+#include "core/framework/data_types.h"
+#include "core/framework/mldata_type_utils.h"
+#include "core/graph/graph_viewer.h"
+#include "core/optimizer/graph_optimizer.h"
+
+namespace onnxruntime {
+
+// Build the MLValue name->idx mapping
+common::Status GraphOptimizer::InitMLValueNameIndexMapping() {
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/graph_optimizer.cc]: >InitMLValueNameIndexMapping [](start = 31, length = 27)

This should be in common code (e.g. helper method in framework utils.h/.cc) and used by both here and SessionStateInitializer::SaveMLValueNameIndexMapping instead of cut-and-pasted.

File: include/onnxruntime/core/framework/op_kernel.h (hunk 1 above)
Comment (weixingzhang) [reply include/onnxruntime/core/framework/op_kernel.h]: Rather than make optimizer depending on ExecutionFrame, I think the better idea is to refactor OpKernelContext and make it not depend on ExecutionFrame. ExecutionFrame was defined for the data of the whole graph. But from kernel point of view, it doesn’t need to know the data of the whole graph. Instead, it just needs to know its inputs, outputs, etc. So, OpKernelContext should be just initialized with the necessary information corresponding to each kernel. The new structure can be created for grouping these necessary information and put it as one of parameters of OpKernelContext ctor.
